File: packages/contracts/src/catalyst_contracts/trace.py
"""Simulation trace contract models (simulation-trace.schema.json).

The trace is the raw, deterministic output of the Rust simulation engine. The
``Portfolio`` / ``PerpPosition`` / ``YieldPosition`` models here are also reused
by the backtest result contract.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import Field

from ._base import Decimal, OpenModel, StrictModel
from .policy import SimulationPolicy
from .request import Interval

EventType = Literal[
    "signal_fired",
    "action_executed",
    "action_rejected",
    "funding_applied",
    "funding_shortfall",
    "yield_accrued",
    "liquidation",
    "gas_charged",
    "fee_charged",
]


class PerpPosition(StrictModel):
    venue: str
    symbol: str
    side: Literal["long", "short"]
    size: Decimal
    entry_price: Decimal
    leverage: Decimal | None = None
    margin_usd: Decimal | None = None
    liquidation_price: Decimal | None = None


class YieldPosition(StrictModel):
    protocol: str
    asset: str
    chain: str
    pool: str | None = None
    principal: Decimal
    accrued: Decimal | None = None


class Portfolio(StrictModel):
    # venue -> asset -> decimal-string balance
    balances: dict[str, dict[str, Decimal]] = Field(default_factory=dict)
    perp_positions: list[PerpPosition] = Field(default_factory=list)
    yield_positions: list[YieldPosition] = Field(default_factory=list)


class Snapshot(StrictModel):
    ts: datetime
    equity_usd: Decimal
    portfolio: Portfolio | None = None


class Event(OpenModel):
    ts: datetime
    type: EventType
    node_id: str | None = None
    reason: str | None = None
    detail: dict[str, Any] | None = None


class SimulationTrace(StrictModel):
    schema_version: str = "catalyst.backtest.trace.v1"
    policy: SimulationPolicy
    interval: Interval
    start: datetime
    end: datetime
    effective_start: datetime | None = None
    effective_end: datetime | None = None
    snapshots: list[Snapshot] = Field(default_factory=list)
    events: list[Event] = Field(default_factory=list)
    final_portfolio: Portfolio
    warnings: list[str] = Field(default_factory=list)
    errors: list[str] = Field(default_factory=list)


__all__ = [
    "SimulationTrace",
    "Snapshot",
    "Event",
    "EventType",
    "Portfolio",
    "PerpPosition",
    "YieldPosition",
]

File: crates/contracts/src/trace.rs
//! Simulation trace contract (simulation-trace.schema.json).

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::policy::SimulationPolicy;
use crate::Decimal;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PerpPosition {
    pub venue: String,
    pub symbol: String,
    pub side: String,
    pub size: Decimal,
    pub entry_price: Decimal,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub leverage: Option<Decimal>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub margin_usd: Option<Decimal>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub liquidation_price: Option<Decimal>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct YieldPosition {
    pub protocol: String,
    pub asset: String,
    pub chain: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub pool: Option<String>,
    pub principal: Decimal,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub accrued: Option<Decimal>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct Portfolio {
    /// venue -> asset -> decimal-string balance
    #[serde(default)]
    pub balances: BTreeMap<String, BTreeMap<String, Decimal>>,
    #[serde(default)]
    pub perp_positions: Vec<PerpPosition>,
    #[serde(default)]
    pub yield_positions: Vec<YieldPosition>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Snapshot {
    pub ts: String,
    pub equity_usd: Decimal,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub portfolio: Option<Portfolio>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Event {
    pub ts: String,
    #[serde(rename = "type")]
    pub event_type: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub node_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub detail: Option<Value>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SimulationTrace {
    #[serde(default = "default_trace_schema_version")]
    pub schema_version: String,
    pub policy: SimulationPolicy,
    pub interval: String,
    pub start: String,
    pub end: String,
    /// First actual tick of the run (#167): the tick clock is data-driven, so
    /// this may be later than the requested `start`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub effective_start: Option<String>,
    /// Last actual tick of the run (#167); may be earlier than `end`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub effective_end: Option<String>,
    #[serde(default)]
    pub snapshots: Vec<Snapshot>,
    #[serde(default)]
    pub events: Vec<Event>,
    pub final_portfolio: Portfolio,
    #[serde(default)]
    pub warnings: Vec<String>,
    #[serde(default)]
    pub errors: Vec<String>,
}

fn default_trace_schema_version() -> String {
    "catalyst.backtest.trace.v1".to_string()
}
